practices/appeng/adrs/0002-distributed-tracing.md

```diff
@@ -0,0 +1,157 @@
+- Distributed Tracing
+  - helpful for learning about unknown unknowns
+  - e.g. "What was the execution context for User X that caused their
+    interaction to timeout last Tuesday?"
+
+Some of the benefits of distributed tracing as outlined in
+[*How Distributed Tracing Solves One of the Worst Microservices Problems*](https://petabridge.com/blog/why-use-distributed-tracing/)
+are:
+
+- radically improves developer productivity and output
+- works across multiple applications, programming languages, and transports
+- improve time to market
+- facilitates excellent cross-team communication and cooperation
+
+Here are several example scenarios or questions that distributed tracing can
+help answer:
+
+- As a new engineer on the team, I want to understand how many separate systems
+  are involved when a certain user type logs in and the first page is rendered.
+- As an operations engineer, I want to know how many SQL queries are executed
+  for a given endpoint or interaction.
+- As a product manager, I want to know if a new feature is being used by a
+  certain cohort of users on a regular basis.
+- As an engineer, I want to prove that an optimization I wrote is effective
+  in a production environment.
+- As a load tester, after I have shown that a problem exists, I want to
+  understand how the system is interacting so I can debug and fix the issue.
+
+### ADR Goals and Anti-goals
+
+- Goal: Choose which set of libraries to use within a codebase (across
+  programming languages) for generating distributed tracing data
+- Anti-goal: Committing to a specific "backend", i.e. platform or service for
+  gathering, exploring, and displaying trace information
+  - Sub-goal: Leaving open as many options as possible for the backend
+
+## Decision
+
+### Use OpenTelemetry
+
+- OpenTelemetry is an emerging industry standard
+  - vendors find benefit of being in the OpenTelemetry ecosystem because they
+    no longer have to create or support instrumentation libraries in an ever
+    growing array of languages, i.e. as soon as language library exists for
+    OpenTelemetry, the vendors automatically become available to support that
+    given language.
+- OpenTelemetry is vendor agnostic
+  - tracing information can be sent to hosted services (e.g. Honeycomb.io, AWS
+    X-Ray, etc) or self-hosted Open Source implementations (e.g. Zipkin, Jaeger,
+    etc)
+  - if left unconfigured, OpenTelemetry instrumentation calls default to
+    lightweight/noop executions
+- OpenTelemetry has well-maintained libraries for the languages used in the
+  layers of most Truss projects
+  - i.e. Go (back-end); JavaScript (front-end); Python (load testing); etc
+- Easily swappable back-ends
+  - e.g. could choose a local Docker version of OpenZipkin for an all-local
+    development environment
+  - e.g. can use Honeycomb.io in the experimental commercial-cloud hosted
+    environment
+  - e.g. can swap in AWS X-Ray for use in GovCloud hosted environments
+- Cons
```

Comment:
The [opentelemetry-go-contrib/instrumentation](https://github.com/open-telemetry/opentelemetry-go-contrib/tree/main/instrumentation) is not merging pull requests. We've had to create a fork of the [http tracing/metrics](https://github.com/trussworks/otelhttp) part because it doesn't have all the features we need. For go, it seems like we're going to have to maintain the metrics library ourselves.

Reply:
Thanks for the context on the challenges with the go implementation.  I think I'd rather that type of info go outside the ADR in a "best practices" type section rather than bake it into the ADR.  I'm hopeful that it gets better soon.

Reply:
I think the instability of some implementations as a Con I think is a valid thing to attribute in this ADR. One of the Pros in this document are that _OpenTelemetry has well-maintained libraries for the languages used in the
   layers of most Truss projects_ which is opposite to the experience @ahobson mentioned. Maybe a good Con to add is that the specification is still early with pre-1.0 versions of libraries being used.

Reply:
I went ahead and added a note about pre-1.0 libraries as a potential con.